forem/forem
# Part One: Fix Users Stuck in a "Created" but "Unregistered" State

## What type of PR is this? (check all applicable)


## Description
This PR is the first of likely a couple of tiny PRs aimed at fixing a user's registration status upon acceptance of an email invitation. This PR addresses the issue that, upon invite acceptance, some Forem users are still stuck in an "unregistered" state. To accomplish this, this PR does the following:
  

- Adds `#after_accept_path_for` to the `Application::Controller` that redirects new users who accepted a Forem invite via email to the onboarding process after setting a password for the new account

- Adds `update!` to the `Invitations::Controller` to update a user's `registered` status from `false` to `true` upon invitation acceptance and their `registered_at` timestamp to the time that the invite was accepted

- Adds a `respond_with` in place of a hard redirect to `Invitations::Controller` 

For further information on Devise's `devise_invitable` gem, please take a look at the GitHub repo: https://github.com/scambra/devise_invitable

## Related Tickets & Documents
Relates to [issue #133](https://github.com/forem/InternalProjectPlanning/issues/133)

## QA Instructions, Screenshots, Recordings
**To QA this PR, please make sure that you have `admin` privileges and then do the following**:

- _**Pro-Tip**_: Replace lines 15 to 22 the `Admin::Invitations::Controller` (right above the `redirect_to admin_invitations_path`) with the code below to make the invite acceptance link readily available in the view after sending an invite:
```
      user = User.invite!(email: email,
                   name: name,
                   username: username,
                   remote_profile_image_url: Users::ProfileImageGenerator.call,
                   saw_onboarding: false,
                   editor_version: :v2,
                   registered: false)
                   flash[:success] = accept_user_invitation_url(invitation_token: user.raw_invitation_token)
```

- Navigate to `/admin/invitations/new` and invite a new user to the site:
![Screen Shot 2020-10-19 at 12 27 44 PM](https://user-images.githubusercontent.com/32834804/96496500-8f681980-1206-11eb-8aad-75b7b1e2b72c.png)


- Using your console, check that the user, after sending the invite, is `registered: false` and has a `registered_at` timestamp of `nil`:
```
[2] pry(main)> user = User.last
[2] pry(main)> user.registered
=> false
[3] pry(main)> user.registered_at
=> nil
```

- After clicking "Create User," the invite acceptance link with the invitation token in it should appear in the view -- copy the link and in an incognito window, paste the invite link. Alternatively, you can copy the link from the view, sign your current Admin user out of `localhost` and then paste the link in the URL bar if you prefer. Either way, you should, after pasting the link in the URL, be brought to the "Set a password to access your account" page:
![Screen Shot 2020-10-19 at 12 06 28 PM](https://user-images.githubusercontent.com/32834804/96494555-bf61ed80-1203-11eb-895a-c8850a4c5717.png)
![Screen Shot 2020-10-19 at 12 08 49 PM](https://user-images.githubusercontent.com/32834804/96494660-e3bdca00-1203-11eb-9e44-12beeada7306.png)

- Set a password for your newly invited user. After setting a password, you should be immediately thrown into the onboarding process, where your user will have to complete the required onboarding before proceeding to use their account:
![Screen Shot 2020-10-19 at 12 17 48 PM](https://user-images.githubusercontent.com/32834804/96495463-23d17c80-1205-11eb-989f-6eab1986311b.png)

- Using your console, check that the user is `registered: true` and has a `registered_at` timestamp:
```
[2] pry(main)> user = User.last
[2] pry(main)> user.registered
=> true
[3] pry(main)> user.registered_at
=> Mon, 19 Oct 2020 18:17:11 UTC +00:00
```

- Navigate to `/admin/users` and check that your newly invited user now shows under the "All" column, showing that they are officially `registered: true`:
![Screen Shot 2020-10-19 at 12 25 26 PM](https://user-images.githubusercontent.com/32834804/96496354-5c258a80-1206-11eb-92a9-96372797a954.png)

## Added tests?


## Added to documentation?


## [optional] Are there any post-deployment tasks we need to perform?
I don't believe that there should be any post-deployment tasks needed at this time, as I already updated the `registered` statuses and `registered_at` timestamps for any Forem users stuck in a state of "created" but "not registered" this past Wednesday.

## [optional] What gif best describes this PR or how it makes you feel?

![Registered](https://media.giphy.com/media/va8eeJC2DEZXF8A4Nx/giphy.gif)
